CIP-array-builtin/README.md

```diff
@@ -0,0 +1,286 @@
+lists for looking up elements inside a collection.
+
+## Specification
+
+We add the following builtin type: `Array` of kind `Type -> Type` representing
+a one-dimensional array type with indices of type `Integer`. Elements are indexed
+consecutively starting from `0`.
+
+***Note***: Here `Type` is the universe of all _builtin_ types, since we do not consider
+types formed out of applying builtin types to arbitrary types to be inhabitted.
+
+The `Array` builtin type should be implemented with a fixed-size immutable array structure
+that has constant-time lookup.
+
+We add the following builtin functions:
+
+1. `indexArray` of type `forall a . Integer -> Array a -> a`.
+    - It returns the element at the given index in the array, or fails with an error if the
+    index is outside the bounds of the array.
+    - It uses constant time and constant memory.
+2. `lengthOfArray` of type `forall a . Array a -> Integer`
+    - It returns the length of the array.
+    - It uses constant time and constant memory.
+3. `listToArray` of type `forall a . List a -> Array a`
+    - It converts the argument builtin list into a builtin array.
+    - It uses linear time and linear memory.
+
+### Binary serialisation and deserialisation
+
+As with all Plutus Core builtin types, arrays must have a fixed binary representation.
+
+For arrays we have chosen the following representation, based on the one currently
+implemented in the [flat][8] encoding for the [Haskell `Array` type][9].
+
+We define the result of encoding a Plutus Core array of type `forall a . Array a`
+and with length `n`, as follows:
+- Let:
+    - `Ea` be an encoding function for `a`, i.e. it is of type `a -> ByteString`.
+    - `Ei` be an encoding function for `Integer`
+    - `++` be bytestring concatenation
+    - `bin(i)` be the binary representation for any number `i`.
+- The first bits in the resulting sequence will be `Ei(0) ++ Ei(n-1)`, that is the
+encoding of the first index followed by the encoding of the last.
+- As binary, arrays are split into _blocks_ of `255` elements. For each block, the first
+bits in the block encoding can be decided as follows:
+    - there will be `n div 255` blocks where the next bits are `bin(255)`
+    - and, if `n mod 255 > 0`, one final block where the next bits are `bin(n mod 255)`. 
+- Inside each block, the next bits come from encoding all the elements pertaining to
+the block:
+    - blocks are indexed starting with `0`;
+    - `indexArray` is a function with the same behaviour as the proposed `indexArray` builtin,
+    since it is fixed, we omit the array argument for brevity;
+    - for the block with size `bs` at index `bn`, the bit sequence will be:
+    `Ea(index(bn*255)) ++ ... ++  Ea(index(bn*255 + bs - 1))`.
+- After all blocks have been encoded, the resulting binary sequence is prepended with exactly
+`8` bits of value `0`. 
+
+***Remark***: The decoding of arrays is, of course, the inverse of the encoding function.
+
+<details>
```

Comment:
I'm finding this very difficult to understand, but I think that's partially because it's very hard to work out what a big string of 0s and 1s is supposed to represent.  I tried some experiments in the repl with `Data.Array` and `flatBits` from `Flat.Instances.Test` and I couldn't get the same restults, but I couldn't really understand how the Flat library was producing its answers anyway.  I think that `flatBits` is adding some extra stuff (not sure  what) before and after the thing that you're actually encoding and that's very confusing.

Reply:
Some of my confusion may also be due to the weird encoding of integers (see C.2.3 and C.2.4 in the PLC specification).

Reply:
Some of my confusion is also due to the weird encoding of integers (see C.2.3 and C.2.4 in the PLC specification).

Reply:
What I did to get these results was to run `flatBits` on an `Array` constructed with https://hackage.haskell.org/package/array-0.5.8.0/docs/Data-Array.html#v:array. I used Haskell's `Integer` type for the indices. In examples 1. and 2. the elements are also `Integer`. In the rest of the examples, the elements are `Bool` (again, Haskell).

Reply:
It turned out that I was looking at `flatBits` of something that was encoded twice, so there was some extra irrelevant stuff  (Sorry! It was late at night).  I'm getting the same answers as it has here now, so it's making more sense.

Reply:
> I was wary of doing that because I think CIPs are supposed to be as accessible as possible. 

Yes, that just make things even more complicated.